### PR TITLE
Fix LocalCUDACluster data spilling and its test

### PR DESCRIPTION
**Pull request by pentschev**
There are two issues fixed here:

1. Creating the local workspace doesn't work, which gets destroyed and recreated by `Worker`;
2. It turns out that the test for `LocalCUDACluster` wasn't running at all, since the internal function didn't get called, and it had some other issues, such as the missing usage of `loop` and handling the proper closure of objects.

### CODE REVIEW

File: dask_cuda/tests/test_spill.py [hunk 1]
```diff
@@ -147,3 +154,13 @@ def get_data(worker, total_size):
                 assert dc > 0
             else:
                 assert dc == 0
+
+        yield client.close()
+
+        # Ensure timeout is respected for workers
+        for w in cluster.workers:
+            yield w.close(timeout=10)
```
Comment (mrocklin) [inline dask_cuda/tests/test_spill.py]: Why was this necessary?  Is `yield cluster.close()` not enough?

File: dask_cuda/tests/test_spill.py [hunk 2]
```diff
@@ -105,16 +108,20 @@ def test_device_spill(client, scheduler, worker):
...
-    @gen_test()
-    def test_cluster_device_spill():
+def test_cluster_device_spill(params):
+    @gen_test(timeout=30)
+    def test():
+        loop = IOLoop.current()
+
         cluster = yield LocalCUDACluster(
             1,
+            loop=loop,
```
Comment (mrocklin) [inline dask_cuda/tests/test_spill.py]: I would expect this to use `IOLoop.current()` by default if asynchronous=True.  Is this not the case?

File: dask_cuda/tests/test_spill.py (hunk 1 above)
Comment (pentschev) [reply dask_cuda/tests/test_spill.py]: No, for some reason the timeout isn't respected, and I had identified where this wasn't working but now why. Now I realized that `LocalCluster` was changed significantly in distributed master, so maybe that would work now. Since we're still on 1.28.1 here, I would keep this and raise an issue to verify this with the new distributed release (1.29?), do you agree with that?

File: dask_cuda/tests/test_spill.py (hunk 2 above)
Comment (pentschev) [reply dask_cuda/tests/test_spill.py]: You're right, we don't need it, at some point I was probably doing something wrong and it worked with this, I'm removing it now.

File: dask_cuda/tests/test_spill.py (hunk 1 above)
Comment (mrocklin) [reply dask_cuda/tests/test_spill.py]: I plan to change our testing here to point to distributed master pretty soon, so we'll have to handle that then.  Lets wait a day for that to happen and then check in here again.

Comment (pentschev) [reply dask_cuda/tests/test_spill.py]: Ok, I will try doing that tomorrow then.

**Comment by pentschev** [comment]
@mrocklin when you have some time to take a look at this, please feel free to comment or merge it if there's no other concerns here.

**Comment by pentschev** [comment]
@mrocklin when you have some time to take a look at this, please feel free to comment or merge it if there's no other concerns here.

File: dask_cuda/tests/test_spill.py [hunk 3]
```diff
@@ -106,44 +108,49 @@ def test_device_spill(client, scheduler, worker):
...
+                await client.run(get_data, cluster.workers[0].Worker, x.nbytes)
+                disk_chunks = await client.run(lambda: len(get_worker().data.disk))
+                for dc in disk_chunks.values():
+                    if params["spills_to_disk"]:
+                        assert dc > 0
+                    else:
+                        assert dc == 0
+
+                await client.close()
+                await cluster.close()
```
Comment (mrocklin) [inline dask_cuda/tests/test_spill.py]:
```suggestion
```

File: dask_cuda/tests/test_spill.py [hunk 4]
```diff
@@ -106,44 +108,49 @@ def test_device_spill(client, scheduler, worker):
...
+
+                await client.run(get_data, cluster.workers[0].Worker, x.nbytes)
+                disk_chunks = await client.run(lambda: len(get_worker().data.disk))
+                for dc in disk_chunks.values():
+                    if params["spills_to_disk"]:
+                        assert dc > 0
+                    else:
+                        assert dc == 0
+
+                await client.close()
```
Comment (mrocklin) [inline dask_cuda/tests/test_spill.py]:
```suggestion
```

These are closed as part of the `async with` context manager

File: dask_cuda/tests/test_spill.py [hunk 5]
```diff
@@ -106,44 +108,49 @@ def test_device_spill(client, scheduler, worker):
...
+                await cluster.close()
 
-        yield client.run(get_data, cluster.workers[0].Worker, x.nbytes)
-        disk_chunks = yield client.run(lambda: len(get_worker().data.disk))
-        for dc in disk_chunks.values():
-            if params["spills_to_disk"]:
-                assert dc > 0
-            else:
-                assert dc == 0
+    loop.run_sync(test)
```
Comment (mrocklin) [inline dask_cuda/tests/test_spill.py]: The other option is to mark the test function as follows:

```python
@pytest.mark.asyncio
async def test_foo(...):
    normal async code here
```

This manages setting up the IOLoop and running it for you.   What we have here is good too though, and also fairly common.

Comment (pentschev) [reply dask_cuda/tests/test_spill.py]: This requires pytest-asyncio or some similar package, is this generally fine to be added?

**Comment by mrocklin** [comment]
Yes

**Comment by pentschev** [comment]
Thanks for the review and suggestions @mrocklin, merging.
